tensorflow_datasets/structured/heart_disease.py

```diff
@@ -0,0 +1,176 @@
+             'non-anginal pain', 'asymptomatic']
+_SLOPE_NAMES = ['upsloping', 'flat', 'downsloping']
+_THAL_NAMES = ['normal', 'fixed defect', 'reversable defect']
+
+_DOWNLOAD_URL = 'http://archive.ics.uci.edu/ml/machine-learning-databases/heart-disease/processed.cleveland.data'
+
+def convert_to_label(d, dictionary):
+  return dictionary[d]
+
+def convert_to_int(d):
+  return int(float(d))
+
+def convert_to_float(d):
+  return float(d)
+
+def return_same(d):
+  return d
+
+FEATURE_DICT = collections.OrderedDict([
+    ("age", (tf.int32, convert_to_int)),
+    ("sex", (tfds.features.ClassLabel(names=['female', 'male']),
+             lambda d: convert_to_label(d, _SEX_DICT))),
+    ("cp", (tfds.features.ClassLabel(names=_CP_NAMES),
+            lambda d: convert_to_label(d, _CP_DICT))),
+    ("trestbps", (tf.int32, convert_to_int)),
+    ("chol", (tf.int32, convert_to_int)),
+    ("fbs", (tfds.features.ClassLabel(names=['false', 'true']),
+             lambda d: convert_to_label(d, _FBS_DICT))),
+    ("restecg", (tf.int32, convert_to_int)),
+    ("thalach", (tf.int32, convert_to_int)),
+    ("exang", (tfds.features.ClassLabel(names=['no', 'yes']),
+               lambda d: convert_to_label(d, _EXE_DICT))),
+    ("oldpeak", (tf.float32, convert_to_float)),
+    ("slope", (tfds.features.ClassLabel(names=_SLOPE_NAMES),
+               lambda d: convert_to_label(d, _SLOPE_DICT))),
+    ("ca", (tf.int32, convert_to_int)),
+    ("thal", (tfds.features.ClassLabel(names=_THAL_NAMES),
+              lambda d: convert_to_label(d, _THAL_DICT)))
+])
+
+
+class HeartDisease(tfds.core.GeneratorBasedBuilder):
+  """Heart disease dataset with 13 attributes."""
+
+  VERSION = tfds.core.Version("0.0.1", "New split API (https://tensorflow.org/datasets/splits)")
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({
+            "features": {name: dtype
+                         for name, (dtype, func) in FEATURE_DICT.items()},
+            "label": tfds.features.ClassLabel(names=['0', '1', '2', '3', '4'])
+        }),
+        supervised_keys=("features", "label"),
+        homepage='http://archive.ics.uci.edu/ml/datasets/Heart+Disease',
+        citation=_CITATION,
+    )
+
+  def _split_generators(self, dl_manager):
+    """Returns SplitGenerators."""
+
+    filepath = dl_manager.download(_DOWNLOAD_URL)
+
+    # There is no predefined train/val/test split for this dataset.
+    return [
+        tfds.core.SplitGenerator(
+            name=tfds.Split.TRAIN,
+            gen_kwargs={"filepath": filepath}
+            ),
+        ]
+
+  def _generate_examples(self, filepath):
+    """Yields examples."""
+
+    feature_columns = ["age", "sex", "cp", "trestbps", "chol", "fbs", "restecg",
+                       "thalach", "exang", "oldpeak", "slope", "ca", "thal"]
+
+    with tf.io.gfile.GFile(filepath) as f:
+      all_lines = f.read().splitlines()
```

Comment:
Would this work with

`csv.DictReader(f, fieldnames=['age',...])`

Reply:
I have implemented this without using csv
```python 
   with tf.io.gfile.GFile(filepath) as f:
      fieldnames = ["age", "sex", "cp", "trestbps",...]
      lines = f.read().splitlines()
      records = [line.replace('?', '-1') for line in lines]
      for i, features in enumerate(records):
        values = {k: float(v) for k, v in zip(fieldnames, features.split(','))}
        yield i, {
            "features": {
                "age": int(values['age']),
                "sex": int(values['sex']),
                "cp": int(values['cp'])-1,
                .....
```
but, we can do something like this with `csv`
```python
     with tf.io.gfile.GFile(filepath) as f:
        fieldnames = ["age", "sex", "cp", "trestbps", ...]
        lines = f.read().splitlines()
        # replace '?' with '-1' 
        records = [line.replace('?', '-1') for line in lines] 
        reader = csv.DictReader(records, fieldnames)
        for i, features in enumerate(reader):
            yield i, {
                "features": {
                    "age": int(float(features['age'])),
                    "sex": int(float(features['sex'])),
                    "cp": int(float(features['cp']))-1,
                    .....
```
We can use both the snippets above, the only thing that bothers me is, we have to import external `csv module`.
Let me know which one is better